app/src/main/java/com/github/geohunt/app/database/Databases.java

```diff
@@ -0,0 +1,30 @@
+package com.github.geohunt.app.database;
+
+import com.google.firebase.database.FirebaseDatabase;
+
+public class Databases {
+    private Databases() {}
+
+    private static Database database = null;
+
+    /**
+     * Retrieve the instance of the database used throughout the program
+     *
+     * @return The database to be used by the application
+     */
+    public static Database getInstance() {
+        if (database == null) {
+            database = new FirebaseDatabaseWrapper(FirebaseDatabase.getInstance());
+        }
+        return database;
+    }
+
+    /**
+     * Set the instance of the database to be used (for testing purpose)
+     *
+     * @param db the database to be used
+     */
+    public static void setInstance(Database db) {
+        database = db;
+    }

```

Comment:
This method is dangerous in my opinion: it totally bypasses the protections given by the private constructor and the singleton instance. A better option for testing purposes would be to mock the static `getInstance` method. This could be done e.g. in Mockito ([https://www.baeldung.com/java-mockito-singleton](url)), and I believe that other mocking frameworks could provide similar effects.

Reply:
1. The private constructor is not a restriction, it is just because it is a utility class
2. I believe that setting the instance should be a public method, not only for testing. I will make it thread-safe though. The only issue that may arise is if the instance is changed, when the old one is still in service however this ain't an issue as the database object does not really have an internal state. It can only perform get/set requests

Reply:
The private constructor actually makes it harder to create database instances that are not the instance provided by `getInstance`, by removing the default constructor. I believe that you could create new instances through reflection, and using this method on such artificial instances could affect the whole application behavior.

I know that the Reflection's API is intended to be verbose to discourage such errors (or malicious usages), but anyway I feel that exposing such a test method to the public would be at best a hack, and at worst an ugly abstraction leak. For testing, mocking framework hide the reflection, so that they are still the best option in terms of usability.

Regarding the possibility of changing the database instance, I do not think that it would be done during the application's runtime, so there is no need to worry about thread-safety.